nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+        .datastore()
+        .get_allocated_regions(disk.volume_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    for (dataset, region) in datasets_and_regions {
+        // Create a Crucible agent client
+        let url = format!("http://{}", dataset.address());
+        let client = CrucibleAgentClient::new(&url);
+
+        // Delete snapshot, it was created by this saga
+        warn!(log, "deleting snapshot {} {} {}", url, region.id(), snapshot_id);
+        client
+            .region_delete_snapshot(
+                &RegionId(region.id().to_string()),
+                &snapshot_id.to_string(),
+            )
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
+    }
+
+    Ok(())
+}
+
+async fn ssc_start_running_snapshot(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<BTreeMap<String, String>, ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // For each dataset and region that makes up the disk, create a map from the
+    // region information to the new running snapshot information.
+    let datasets_and_regions = osagactx
+        .datastore()
+        .get_allocated_regions(disk.volume_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let mut map: BTreeMap<String, String> = BTreeMap::new();
+
+    for (dataset, region) in datasets_and_regions {
+        // Create a Crucible agent client
+        let url = format!("http://{}", dataset.address());
+        let client = CrucibleAgentClient::new(&url);
+
+        warn!(log, "dataset {:?} region {:?} url {}", dataset, region, url);
+
+        // Validate with the Crucible agent that the snapshot exists
+        let crucible_region = client
+            .region_get(&RegionId(region.id().to_string()))
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
+
+        warn!(log, "crucible region {:?}", crucible_region);
+
+        let crucible_snapshot = client
+            .region_get_snapshot(
+                &RegionId(region.id().to_string()),
+                &snapshot_id.to_string(),
+            )
+            .await
+            .map_err(|e| e.to_string())
+            .map_err(ActionError::action_failed)?;
+
+        warn!(log, "crucible snapshot {:?}", crucible_snapshot);
+
+        // Start the snapshot running
+        let crucible_running_snapshot = client
+            .region_run_snapshot(
```

Comment:
Confirming: This endpoint is idempotent, correct? It needs to be for this saga node to be idempotent.

In the simulated crucible agent implementation, this is defined as a POST, which would suggest otherwise.

Reply:
This endpoint (along with other crucible agent ones) should be idempotent, yes. [The path that creates read-write downstairs is](https://github.com/oxidecomputer/crucible/blob/0f18733ba0136e01d686812893d7a746c1c882d1/agent/src/datafile.rs#L194-L198), but [the path that creates read-only downstairs is not](https://github.com/oxidecomputer/crucible/blob/0f18733ba0136e01d686812893d7a746c1c882d1/agent/src/datafile.rs#L249-L251). The Nexus code should stay the same, and [the Crucible agent impl has to change](https://github.com/oxidecomputer/crucible/issues/380).